### PR TITLE
Issue #2073 from the AS repository where due to the ruamel update to …

### PR DESCRIPTION
**Pull request by VindeeR**
First part of the solution for the issue https://github.com/BSC-ES/autosubmit/issues/2073 from the AS repository where due to the ruamel update to the version 0.18 A deprecated version of dump function was generating a warning when cloning a project from AS3 to the new AS4

### CODE REVIEW

**Review by kinow — COMMENTED** [review]
The change is small, but looks like GitHub Actions is not happy with this. Check the GH Actions logs to see if you can fix it :slightly_smiling_face:

**Review by kinow — COMMENTED** [review]
Looking at the Codecov missing lines, it seems like the code you changed was not tested before, @VindeeR . So it's good practice to write a unit test to test that the code does what's expected, and to prevent the error from happening again in the future.

image

You can have a look at existing unit tests, at the [pytest docs](https://docs.pytest.org/en/stable/), and see if you can write a test for that. Normally I write a test that will fail before the change. Then run the test with the change (in this case, you can just comment/comment out the code), and then commit and push.

Once the test runs in the GH Actions, then the coverage tool should detect this lines of code were "covered" by your new test, marking the build green/passed :+1:

**Review by kinow — CHANGES_REQUESTED** [review]
Thanks for adding the test in the last commits, @VindeeR !

I think we can cover the changes with a simpler test, a unit test (although it'd be nice to have a more real functional test -- we have some issues in AS for that).

Here's what I did just now:

- Sync code with Git
- Open IDE with your branch
- Ran your test, and saw the error
- Undid your changes (since they were small, and I wanted to see the error)
- Started throwing away parts of the code I wasn't going to use in this test, and added other test to just call the function you changed (added some type hints here and there where code was being tested/called, and also found a :bug: in AS https://github.com/BSC-ES/autosubmit/issues/2085)

```diff
diff --git a/test/unit/conftest.py b/test/unit/conftest.py
index 8a68797..a7db17f 100644
--- a/test/unit/conftest.py
+++ b/test/unit/conftest.py
@@ -217,7 +217,8 @@ def autosubmit_config(
 
     def finalizer() -> None:
         BasicConfig.LOCAL_ROOT_DIR = original_root_dir
-        rmtree(tmp_path)
+        if tmp_path and tmp_path.exists():
+            rmtree(tmp_path)
 
     request.addfinalizer(finalizer)
 
diff --git a/test/unit/test_configcommon.py b/test/unit/test_configcommon.py
index 230a79e..9f1c98a 100644
--- a/test/unit/test_configcommon.py
+++ b/test/unit/test_configcommon.py
@@ -127,15 +127,15 @@ def test_clean_dynamic_variables(autosubmit_config: Callable) -> None:
 
 
 def test_yaml_deprecation_warning(autosubmit_config: Callable):
-
-    as_conf: AutosubmitConfig = autosubmit_config(expid='a000', experiment_data={})
-    as_conf_copy: AutosubmitConfig = autosubmit_config(expid='a001', experiment_data={})
-
-    as_path = Path.home() / 'autosubmit'
-    expid = str(as_conf.expid)
-    expid_copy = str(as_conf_copy.expid)
-    file_base = "conf/expdef_"+expid+".conf"
-    file_copied = "conf/expdef_"+expid_copy+".conf_AS_v3_backup"
-    expected = open(as_path/expid_copy/file_copied, 'r')
-    with open(as_path/expid/file_base, 'r') as file:
-        assert expected.read() == file.read()
\ No newline at end of file
+    pass
+    # as_conf: AutosubmitConfig = autosubmit_config(expid='a000', experiment_data={})
+    # as_conf_copy: AutosubmitConfig = autosubmit_config(expid='a001', experiment_data={})
+    #
+    # as_path = Path.home() / 'autosubmit'
+    # expid = str(as_conf.expid)
+    # expid_copy = str(as_conf_copy.expid)
+    # file_base = "conf/expdef_"+expid+".conf"
+    # file_copied = "conf/expdef_"+expid_copy+".conf_AS_v3_backup"
+    # expected = open(as_path/expid_copy/file_copied, 'r')
+    # with open(as_path/expid/file_base, 'r') as file:
+    #     assert e
```
- Then ran the test in my IDE, and got the error that I think originated the issue:

```bash
/home/kinow/Development/python/micromamba/envs/autosubmit4/bin/python /home/kinow/Development/python/pycharm-2024.1.4/plugins/python/helpers/pycharm/_jb_pytest_runner.py --target unit/test_configcommon.py::test_yaml_deprecation_warning 
Testing started at 11:17 AM ...
Launching pytest with arguments unit/test_configcommon.py::test_yaml_deprecation_warning --no-header --no-summary -q in /home/kinow/Development/python/workspace/autosubmit-config-parser/test

============================= test session starts ==============================
collecting ... collected 1 item

unit/test_configcommon.py::test_yaml_deprecation_warning 

======================== 1 failed, 4 warnings in 0.28s =========================
FAILED          [100%]Backup stored at /tmp/pytest-of-kinow/pytest-11/test_yaml_deprecation_warning0/a000_jobs.ini_AS_v3_backup

test/unit/test_configcommon.py:129 (test_yaml_deprecation_warning)
tmp_path = PosixPath('/tmp/pytest-of-kinow/pytest-11/test_yaml_deprecation_warning0')
autosubmit_config = <function autosubmit_config.<locals>._create_autosubmit_config at 0x70115748c4c0>

    def test_yaml_deprecation_warning(tmp_path, autosubmit_config: Callable):
        """Test that the conversion from YAML to INI works as expected, without warnings.
    
        Creates a dummy AS3 INI file, calls ``AutosubmitConfig.ini_to_yaml``, and
        verifies that the YAML files exists and is not empty, and a backup file was
        created. All without warnings being raised (i.e. they were suppressed).
        """
        as_conf: AutosubmitConfig = autosubmit_config(expid='a000', experiment_data={})
    
        ini_file = tmp_path / 'a000_jobs.ini'
        with open(ini_file, 'w+') as f:
            f.write(dedent('''\
            [LOCAL_SETUP]
            FILE = LOCAL_SETUP.sh
            PLATFORM = LOCAL
            '''))
            f.flush()
    
>       as_conf.ini_to_yaml(root_dir=tmp_path, ini_file=str(ini_file))

unit/test_configcommon.py:148: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../autosubmitconfigparser/config/configcommon.py:2559: in ini_to_yaml
    yaml.dump(final_dict, yaml_file, Dumper=yaml.RoundTripDumper)
../../../micromamba/envs/autosubmit4/lib/python3.9/site-packages/ruamel/yaml/main.py:1251: in dump
    error_deprecation('dump', 'dump', arg="typ='unsafe', pure=True")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fun = 'dump', method = 'dump', arg = "typ='unsafe', pure=True"
comment = 'instead of'

    def error_deprecation(fun: Any, method: Any, arg: str = '', comment: str = 'instead of') -> None:  # NOQA
        import inspect
    
        s = f'\n"{fun}()" has been removed, use\n\n  yaml = YAML({arg})\n  yaml.{method}(...)\n\n{comment}'  # NOQA
        try:
            info = inspect.getframeinfo(inspect.stack()[2][0])
            context = '' if info.code_context is None else "".join(info.code_context)
            s += f' file "{info.filename}", line {info.lineno}\n\n{context}'
        except Exception as e:
            _ = e
        s += '\n'
        if sys.version_info < (3, 10):
>           raise AttributeError(s)
E           AttributeError: 
E           "dump()" has been removed, use
E           
E             yaml = YAML(typ='unsafe', pure=True)
E             yaml.dump(...)
E           
E           instead of file "/home/kinow/Development/python/workspace/autosubmit-config-parser/autosubmitconfigparser/config/configcommon.py", line 2559
E           
E                   yaml.dump(final_dict, yaml_file, Dumper=yaml.RoundTripDumper)

../../../micromamba/envs/autosubmit4/lib/python3.9/site-packages/ruamel/yaml/main.py:1037: AttributeError

Process finished with exit code 1
```
- Removed my changes in `configcommon.py`, leaving your fix back in place instead;
- Ran the test again and with coverage (PyCharm has an option to do that, you can use other IDEs, editors, pytest+pytest-cov, etc.):

image

- To finish, I ran all tests just to make sure I was not breaking anything, simply running `pytest`

And bam, :boom: , another bug :bug: :rofl: 

```bash
autosubmitconfigparser/config/configcommon.py:2548: in ini_to_yaml
    yaml_dict = update_dict(yaml_dict, value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

original_dict = {}, updated_dict = {'LOCAL_SETUP': {'FILE': 'LOCAL_SETUP.sh', 'PLATFORM': 'LOCAL'}}

    def update_dict(original_dict: Dict, updated_dict: Dict) -> Dict:
        for k, v in updated_dict.items():
>           if isinstance(v, collections.Mapping):
E           AttributeError: module 'collections' has no attribute 'Mapping'

autosubmitconfigparser/config/configcommon.py:2508: AttributeError
```

Funny that it passed running the test individually in PyCharm (the IDE does some magic to capture stdout/stderr, and launch the tests differently, so it might be that).

Turns out our code has `collection.Mapping` in the code you changed, but that would fail later on as the correct class is `collection.abc.Mapping`.

Here's the final diff, @VindeeR . Take your time to review this review (:drum:), let me know if you have any questions, and in theory following more or less what I did should give you a green build with your code covered by the new unit test your wrote.

```diff
diff --git a/autosubmitconfigparser/config/configcommon.py b/autosubmitconfigparser/config/configcommon.py
index 8cb92a9..a395522 100644
--- a/autosubmitconfigparser/config/configcommon.py
+++ b/autosubmitconfigparser/config/configcommon.py
@@ -2501,11 +2501,11 @@ class AutosubmitConfig(object):
 
     # based on https://github.com/cbirajdar/properties-to-yaml-converter/blob/master/properties_to_yaml.py
     @staticmethod
-    def ini_to_yaml(root_dir, ini_file):
+    def ini_to_yaml(root_dir: Path, ini_file: str) -> None:
         # Based on http://stackoverflow.com/a/3233356
-        def update_dict(original_dict, updated_dict):
+        def update_dict(original_dict: Dict, updated_dict: collections.abc.Mapping) -> Dict:
             for k, v in updated_dict.items():
-                if isinstance(v, collections.Mapping):
+                if isinstance(v, collections.abc.Mapping):
                     r = update_dict(original_dict.get(k, {}), v)
                     original_dict[k] = r
                 else:
diff --git a/test/unit/conftest.py b/test/unit/conftest.py
index 8a68797..a7db17f 100644
--- a/test/unit/conftest.py
+++ b/test/unit/conftest.py
@@ -217,7 +217,8 @@ def autosubmit_config(
 
     def finalizer() -> None:
         BasicConfig.LOCAL_ROOT_DIR = original_root_dir
-        rmtree(tmp_path)
+        if tmp_path and tmp_path.exists():
+            rmtree(tmp_path)
 
     request.addfinalizer(finalizer)
 
diff --git a/test/unit/test_configcommon.py b/test/unit/test_configcommon.py
index 230a79e..3584adf 100644
--- a/test/unit/test_configcommon.py
+++ b/test/unit/test_configcommon.py
@@ -1,5 +1,6 @@
 from typing import Callable
 from pathlib import Path
+from textwrap import dedent
 import pytest
 
 from autosubmitconfigparser.config.configcommon import AutosubmitConfig
@@ -126,16 +127,22 @@ def test_clean_dynamic_variables(autosubmit_config: Callable) -> None:
     assert 'jaspion_eats' in as_conf.dynamic_variables
 
 
-def test_yaml_deprecation_warning(autosubmit_config: Callable):
+def test_yaml_deprecation_warning(tmp_path, autosubmit_config: Callable):
+    """Test that the conversion from YAML to INI works as expected, without warnings.
 
+    Creates a dummy AS3 INI file, calls ``AutosubmitConfig.ini_to_yaml``, and
+    verifies that the YAML files exists and is not empty, and a backup file was
+    created. All without warnings being raised (i.e. they were suppressed).
+    """
     as_conf: AutosubmitConfig = autosubmit_config(expid='a000', experiment_data={})
-    as_conf_copy: AutosubmitConfig = autosubmit_config(expid='a001', experiment_data={})
-
-    as_path = Path.home() / 'autosubmit'
-    expid = str(as_conf.expid)
-    expid_copy = str(as_conf_copy.expid)
-    file_base = "conf/expdef_"+expid+".conf"
-    file_copied = "conf/expdef_"+expid_copy+".conf_AS_v3_backup"
-    expected = open(as_path/expid_copy/file_copied, 'r')
-    with open(as_path/expid/file_base, 'r') as file:
-        assert expected.read() == file.read()
\ No newline at end of file
+
+    ini_file = tmp_path / 'a000_jobs.ini'
+    with open(ini_file, 'w+') as f:
+        f.write(dedent('''\
+        [LOCAL_SETUP]
+        FILE = LOCAL_SETUP.sh
+        PLATFORM = LOCAL
+        '''))
+        f.flush()
+
+    as_conf.ini_to_yaml(root_dir=tmp_path, ini_file=str(ini_file))
```

Cheers

File: test/unit/test_configcommon.py [hunk 1]
```diff
@@ -124,3 +125,38 @@ def test_clean_dynamic_variables(autosubmit_config: Callable) -> None:
...
     assert 'jaspion_eats' in as_conf.dynamic_variables
+
+
+def test_yaml_deprecation_warning(tmp_path, autosubmit_config: Callable):
+    """Test that the conversion from YAML to INI works as expected, without warnings.
+
+    +    Creates a dummy AS3 INI file, calls ``AutosubmitConfig.ini_to_yaml``, and
+    +    verifies that the YAML files exists and is not empty, and a backup file was
+    +    created. All without warnings being raised (i.e. they were suppressed).
+    +    """
```
Comment (kinow) [inline test/unit/test_configcommon.py:136]: These `+` signs are probably from the Git diff :grimacing: They need to be removed...

File: test/unit/test_configcommon.py [hunk 2]
```diff
@@ -124,3 +125,38 @@ def test_clean_dynamic_variables(autosubmit_config: Callable) -> None:
...
+            FILE = LOCAL_SETUP.sh
+            PLATFORM = LOCAL
+            '''))
+        f.flush()
+    as_conf.ini_to_yaml(root_dir=tmp_path, ini_file=str(ini_file))
+
+    print(f'tmp_path: {tmp_path}')
+    with open(tmp_path / 'a000_jobs.ini_AS_v3_backup', 'r') as backup, open(tmp_path / 'a000_jobs.yml', 'r') as newYaml:
+        lines_backup = backup.readlines()
+        lines_newYaml = newYaml.readlines()
```
Comment (kinow) [inline test/unit/test_configcommon.py:152]: Camel case is not used in Python. So never `newYaml`, only snake case like `new_yaml` is used.

Comment (kinow) [reply test/unit/test_configcommon.py:152]: You may want to configure your editor or IDE so issues like this are automatically flagged, before you push to branches. Here's what PyCharm looks like when I open your code.

image

There are light wiggly lines under some parts of the code, that also appear on the right sidebar, and as a summary at the top right of my IDE. It shows 2 warnings for this file (the `False` at the end, and this camel case).

File: test/unit/test_configcommon.py [hunk 3]
```diff
@@ -124,3 +125,38 @@ def test_clean_dynamic_variables(autosubmit_config: Callable) -> None:
...
+    with open(ini_file, 'w+') as f:
+        f.write(dedent('''\
+            [LOCAL_SETUP]
+            FILE = LOCAL_SETUP.sh
+            PLATFORM = LOCAL
+            '''))
+        f.flush()
+    as_conf.ini_to_yaml(root_dir=tmp_path, ini_file=str(ini_file))
+
+    print(f'tmp_path: {tmp_path}')
```
Comment (kinow) [inline test/unit/test_configcommon.py:149]: You should not print in tests. It is bad practice, and in fact pytest won't output it, as pytest captures the stdout/stderr (you can tell it to output, but in some corner cases that doesn't work as expected).

If you use this for writing your tests, just remember to delete it before pushing or before assigning a reviewer.

File: test/unit/test_configcommon.py [hunk 4]
```diff
@@ -124,3 +125,38 @@ def test_clean_dynamic_variables(autosubmit_config: Callable) -> None:
...
+
+        valid = True
+        for line_backup, line_newYaml in zip(lines_backup[1:], lines_newYaml[2:]):
+            if line_backup.strip().split('=')[1] != line_newYaml.strip().split(':')[1]:
+                valid = False
+
+        if valid:
+            assert True
+        elif valid:
+            assert False
```
Comment (kinow) [inline test/unit/test_configcommon.py:162]: I think this test could be simpler... at the moment one needs to follow the lifecycle of that `valid` variable. Checking where it may change value. The test also checks that certain lines of the INI and YAML match, but this could be a bit fragile if someone changes something in the logic (I think we'd need a separate test for that).

Maybe something like this?

```py
def test_yaml_deprecation_warning(tmp_path, autosubmit_config: Callable):
    """Test that the conversion from YAML to INI works as expected, without warnings.

    Creates a dummy AS3 INI file, calls ``AutosubmitConfig.ini_to_yaml``, and
    verifies that the YAML file exists and is not empty, and a backup file was
    created. All without warnings being raised.
    """

    as_conf: AutosubmitConfig = autosubmit_config(expid='a000', experiment_data={})
    ini_file = tmp_path / 'a000_jobs.ini'
    with open(ini_file, 'w+') as f:
        f.write(dedent('''\
            [LOCAL_SETUP]
            FILE = LOCAL_SETUP.sh
            PLATFORM = LOCAL
            '''))
        f.flush()

    as_conf.ini_to_yaml(root_dir=tmp_path, ini_file=str(ini_file))

    backup_file = Path(f'{ini_file}_{_BACKUP_SUFFIX}')
    assert backup_file.exists()
    assert backup_file.stat().st_size > 0

    new_yaml_file = Path(ini_file.parent, ini_file.stem).with_suffix('.yml')
    assert new_yaml_file.exists()
    assert new_yaml_file.stat().st_size > 0
```

**Review by kinow — CHANGES_REQUESTED** [review]

**Review by kinow — APPROVED** [review]
Looks good to me! Please make sure it is squashed
